app/graphql/hooks.ts

```diff
@@ -186,3 +195,114 @@ export const useDataCubesComponentsQuery = makeUseQuery<
   DataCubesComponentsOptions,
   DataCubesComponentsData
 >(executeDataCubesComponentsQuery);
+
+type DataCubesObservationsOptions = {
+  variables: Omit<DataCubeComponentsQueryVariables, "cubeFilter"> & {
+    cubeFilters: DataCubeObservationFilter[];
+  };
+  pause?: boolean;
+};
+
+type DataCubesObservationsData = {
+  dataCubesObservations: DataCubesObservations;
+};
+
+export const executeDataCubesObservationsQuery = async (
+  variables: DataCubesObservationsOptions["variables"],
+  /** Callback triggered when data fetching starts (cache miss). */
+  onFetching?: () => void
+) => {
+  const { locale, sourceType, sourceUrl, cubeFilters } = variables;
+
+  if (cubeFilters.length > 1 && !cubeFilters.every((f) => f.joinBy)) {
+    throw new Error(
+      "When fetching data from multiple cubes, all cube filters must have joinBy property set."
+    );
+  }
+
+  const queries = await Promise.all(
+    cubeFilters.map((cubeFilter) => {
+      const cubeVariables = { locale, sourceType, sourceUrl, cubeFilter };
+      const cached = client.readQuery<
+        DataCubeObservationsQuery,
+        DataCubeObservationsQueryVariables
+      >(DataCubeObservationsDocument, cubeVariables);
+
+      if (cached) {
+        return Promise.resolve(cached);
+      }
+
+      onFetching?.();
+
+      return client
+        .query<DataCubeObservationsQuery, DataCubeObservationsQueryVariables>(
+          DataCubeObservationsDocument,
+          cubeVariables
+        )
+        .toPromise();
+    })
+  );
+
+  const error = queries.find((q) => q.error)?.error;
+  const fetching = !error && queries.some((q) => !q.data);
+
+  if (fetching) {
+    return {
+      data: undefined,
+      error,
+      fetching,
+    };
+  }
+
+  const observations =
+    // If we are fetching data from multiple cubes, we need to merge them into one
+    cubeFilters.length > 1
+      ? queries.reduce<Record<string | number, Observation>>((acc, query) => {
+          const joinBy = query.operation.variables?.cubeFilter.joinBy as string;
+          const observations = query.data?.dataCubeObservations
+            ?.data as Observation[];
+
+          for (const observation of observations) {
+            const key = observation[joinBy];
+
+            if (!key) {
+              continue;
+            }
+
+            const existing = acc[key];
+            acc[key] = Object.assign(existing ?? {}, observation);
+          }
+
+          return acc;
+        }, {})
+      : // If we are fetching data from a single cube, we can just return the data
+        (queries[0].data?.dataCubeObservations?.data as Observation[]);
```

Comment:
after I see that we do Object.values on the observations so it seems like in the case of multiple cubes, observations is an index, while when it is 1 cube, it is only an array ?

Reply:
Yes! In this case we can call Object.values on both types to get what we need (for array it will return the same array) 👍 But maybe it could be a bit more explicit :)

Reply:
Oh I did not know that Object.values was working for arrays. TIL.

Nonetheless, I think it is always good if the type of a variables does not change across code branches, maybe we could call Object.values a bit earlier so that we are sure that observations is always an array ?